apps/arweave/src/ar_http_iface_server.erl

```diff
@@ -95,8 +94,181 @@ start_http_iface_listener(Config) ->
+	% the socket is already down.
+	try inet:peername(Socket) of
+		{ok, {PeerAddress, PeerPort}} ->
+			% let  prepare the state used for the termination loop procedure.
+			Peer = string:join([inet:ntoa(PeerAddress), integer_to_list(PeerPort)], ":"),
+			State = #{
+				socket => Socket,
+				peer => Peer,
+				delay => Delay,
+				connection => Connection
+			},
+
+			% now everything is ready, the socket is set in read-only mode. the
+			% remote peer will receive this information and will try to fetch
+			% the last piece of data from the buffer. It does not guarantee the
+			% socket will be closed, in particular if the connection is slow.
+			logger:warning(#{ connection => Peer, socket => Connection, reason => "shutdown", msg => "shutdown (read-only)"}),
+			ranch_tcp:shutdown(Socket, write),
+
+			% in this case, two timeouts are configured, one simply closing the
+			% socket to announce another time to the remote peer that the node is
+			% shutting down...
+			{ok, _} = timer:send_after(Delay, {timeout, close}),
+
+			% ... and a second delay. This time, the socket is killed. linger
+			% option is set to 0 on the port and we reclose the socket.
+			{ok, _} = timer:send_after(Delay*2, {timeout, kill}),
+
+			% let start the terminate loop to shutdown properly the connection
+			terminate_connection_loop(State);
+		_ ->
+			ok
+	catch
+		_:_ ->
+			ok
+	end.
+
+%%--------------------------------------------------------------------
+%% @hidden
+%% @doc shutdown procedure loop, in charge of dealing with monitoring
+%%      messages and timeouts.
+%% @end
+%%--------------------------------------------------------------------
+-spec terminate_connection_loop(State) -> Return when
+	Socket :: port(),
+	Connection :: pid(),
+	Peer :: string(),
+	Delay :: pos_integer(),
+	State :: #{
+		socket => Socket,
+		connection => Connection,
+		peer => Peer,
+		delay => Delay
+	},
+	Return :: ok.
+
+terminate_connection_loop(State) ->
+	Socket = maps:get(socket, State),
+	Connection = maps:get(connection, State),
+	Peer = maps:get(peer, State),
+	Delay = maps:get(delay, State),
+	AfterDelay = Delay*3,
+	receive
+		{timeout, close} ->
+			% the first timeout simply close the socket a
+			% socket time.
+			logger:warning(#{ connection => Peer, socket => Connection, reason => "shutdown", msg => "closed (timeout)"}),
+			ranch_tcp:close(Socket),
+			terminate_connection_loop(State);
+
+		{timeout, kill} ->
+			% the second timeout set linger tcp parameter
+			% to 0 and close the connection (again). At
+			% this stage, the connection should be closed.
+			% and we should receive a message from
+			% monitoring.
+			logger:warning(#{ connection => Peer, socket => Connection, reason => "shutdown", msg => "killed (timeout)"}),
+			ranch_tcp:setopts(Socket, [{linger, {true, 0}}]),
+			ranch_tcp:close(Socket),
+			terminate_connection_loop(State);
+
+		{'DOWN', _Ref, process, Connection, _} ->
```

Comment:
If this message comes in before one or both of the `{timeout, xxx}` messages, that's okay, right? Nothing else needs to be cleaned up because the connection has already terminated, and the pending timeout messages will just be sent into the void